### PR TITLE
Introduce infrastructure for generating target docs

### PR DESCRIPTION
**Pull request by Noratrieb**
See #120745 

It's highly unlikely that the format is optimal, but it's okay at least and can always be improved. Mostly posting this to get something working so we can continue.

Use `TARGET_CHECK_ONLY=0` to actually build the book instead of just checking the new docs.

r? @davidtwco

### CODE REVIEW

**Comment by Nilstrieb** [comment]
this is a very funny failure, sadly I cannot reproduce it locally.

**Comment by ehuss** [comment]
@Nilstrieb The issue is when it runs `--print cfg --target csky-unknown-linux-gnuabiv2` it fails because LLVM 16 does not support that target.

It tries to print the error "No available targets are compatible with triple \"csky-unknown-linux-gnuabiv2\"".

However, one of the arguments is a SmallCStr which is created from a CStr. I believe there is a bug [here](https://github.com/rust-lang/rust/blame/ee9c7c940c07d8b67c9a6b2ec930db70dcd23a46/compiler/rustc_data_structures/src/small_c_str.rs#L85) where `SmallCStr` is created without the trailing null byte. It then later calls `as_c_str` which does `from_bytes_with_nul_unchecked`, but there is no null.

I can maybe try to post a fix for the panic later tonight. I don't know how best to handle the problem with LLVM 16 not supporting the target. Perhaps `target-docs` could check stderr for the "no available targets" message, and just skip that target?

**Comment by ehuss** [comment]
Posted #121125 so it at least doesn't ICE.

**Review by davidtwco — COMMENTED** [review]
I really like this approach. I think the only feedback I've got is about where some of the specific bits of information go.

This approach shines with the overlapping documentation based with patterns, but if you need to have some information for each target (like tier, maintainers, descriptive name, supports std, etc.) and have to make documentation for every target anyway, then you lose some of that benefit.

I think I'd like to see us experiment with putting some of that information into the target specification, and then have a way of dumping that from the compiler (like you do with the cfgs now).

**Comment by Nilstrieb** [comment]
I like that. We could even print the target description with `--print target-list --verbose` :). I will take a look.

**Comment by Nilstrieb** [comment]
this is blocked on work on the tracking issue

File: src/doc/rustc/src/SUMMARY.md
Comment (Nilstrieb) [inline src/doc/rustc/src/SUMMARY.md:72]: This needs to be removed or commented out before merging.

**Comment by Nilstrieb** [comment]
@rustbot ready
Tidy is failing on purpose, this is ready for another review.

File: src/tools/target-docs/src/parse.rs
Comment (Nilstrieb) [inline src/tools/target-docs/src/parse.rs:103]: This **clearly** needs a dependency on https://crates.io/crates/bool-toggle to make the code clearer

File: src/tools/target-docs/src/render.rs
Comment (jieyouxu) [inline src/tools/target-docs/src/render.rs:91]: These can probably be `const`s

**Review by jieyouxu — COMMENTED** [review]
Left some tiny nits :3

File: src/tools/target-docs/src/render.rs
Comment (Nilstrieb) [reply src/tools/target-docs/src/render.rs:91]: i dont think so

File: src/bootstrap/src/lib.rs
Comment (Nilstrieb) [reply src/bootstrap/src/lib.rs:1734]: maybe, but i just copied it so it doesnt really matter

File: src/tools/target-docs/src/render.rs
Comment (jieyouxu) [reply src/tools/target-docs/src/render.rs:91]: Nvm, for some reason I thought they were just constant slices 🤦

File: src/bootstrap/src/core/build_steps/doc.rs
Comment (ehuss) [reply src/bootstrap/src/core/build_steps/doc.rs:1153]: Using `.join("foo/bar")` is fine on all platforms. Windows generally allows mixing either separator.

File: src/doc/rustc/target_infos/*-apple-tvos.md
Comment (Nilstrieb) [inline src/doc/rustc/target_infos/*-apple-tvos.md:1]: I think it makes sense to put a very big warning comment in the old docs for the docs that have already been migrated, reminding people to update them in both places to keep them in sync for now.

**Review by davidtwco — APPROVED** [review]
LGTM, r=me with the bits that need removed before merging removed, and the TODOs for #122590 resolved. I don't mind if you add a comment to keep the old and new bits in sync, hopefully we can follow up on moving everything fairly soon.

**Comment by Nilstrieb** [comment]
ugh, llvm 16 doesn't support this target :/. that sucks a lot, maybe there should be an env var like `SKIP_TARGETS` env var that don't get documentation generated and then the llvm 16 builder could set that.

**Comment by workingjubilee** [comment]
Odd, I thought CSky had landed with LLVM 15.

cc @Dirreke You're the maintainer for this target, right?

Maybe we should just raise our LLVM to 17...

**Comment by Nilstrieb** [comment]
I excluded the csky targets the (now) LLVM 17 builder.
i think at this point i should also do
r? bootstrap
for all of my bootstrap crimes

File: src/bootstrap/src/lib.rs
Comment (onur-ozkan) [inline src/bootstrap/src/lib.rs:1677]: It's not doing anything non-ordinary than simple copying, so I don't think `cp_slow` is a good choice for naming this. Can we rename this function to `copy` and remove the `copy` part from other function names that are linking files instead of copying? Hopefully this can stop this confusion going forward.

File: src/bootstrap/src/core/build_steps/doc.rs
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/doc.rs:1207]: Can you put the actual link here? It's not clear that whether this is an issue or a PR id.

Comment (Nilstrieb) [reply src/bootstrap/src/core/build_steps/doc.rs:1207]: Issues and PRs share a namespace :)

File: src/bootstrap/src/lib.rs
Comment (Nilstrieb) [reply src/bootstrap/src/lib.rs:1677]: My train of thought here was that everything in bootstrap used `copy_link`, so `copy_link` should probably be the default picked by people and `copy_slow` is only used in one place and shouldn't really be needed, so I'd want to discourage people from using it unless they really need to. Especially given that some people might already know that `copy` existed and reach for it, not realizing that it was changed.

File: src/bootstrap/src/core/build_steps/doc.rs
Comment (onur-ozkan) [reply src/bootstrap/src/core/build_steps/doc.rs:1207]: Interesting, I didn't know that. I would still prefer the actual link though, since most editors will be able to open it by clicking on it.

File: src/bootstrap/src/lib.rs
Comment (onur-ozkan) [reply src/bootstrap/src/lib.rs:1677]: `copy_slow` really seems very confusing and not correct..

Comment (Nilstrieb) [reply src/bootstrap/src/lib.rs:1677]: I can rename it to `copy` if you prefer :+1:

**Review by onur-ozkan — APPROVED** [review]
bootstrap changes looks good to me

**Comment by Nilstrieb** [comment]
thanks for the review!
@bors=davidtwco,onur-ozkan

**Comment by Nilstrieb** [comment]
uh
@bors r=davidtwco,onur-ozkan

**Comment by workingjubilee** [comment]
https://github.com/rust-lang/rust/pull/123010#issuecomment-2017023084 @Nilstrieb I think this was yours, unfortunately.

@bors r-

**Comment by workingjubilee** [comment]
@rollup=iffy

**Comment by Nilstrieb** [comment]
I see, Windows exists. Good point. I changed the filename to use `_` instead of `*` for wildcards. Since that collides with `_` present in some targets (like that one obscure architecture), I instead moved the pattern to the frontmatter, where asterisks are allowed (supposedly). target-docs enforces that the file name is consistent with the pattern.
I didn't rebase, so the force push compare page is accurate.
@rustbot ready

**Review by onur-ozkan — COMMENTED** [review]
Changes in the last push seems solid. Can we add unit test in `parse/tests.rs` for the logic here? https://github.com/rust-lang/rust/blob/1219d99fc95d5673bcb566efae0ab7abc6bc06d3/src/tools/target-docs/src/parse.rs#L87-L93

**Comment by klensy** [comment]
Well, why `glob-match` crate? There already `glob` and `globset` in workspace, and current usage should be doable via https://docs.rs/glob/latest/glob/struct.Pattern.html#method.matches.

**Comment by davidtwco** [comment]
r=me with [this comment resolved](https://github.com/rust-lang/rust/pull/121051#pullrequestreview-1968497756) and if it isn't hard to use `glob` or `globset` instead, then that would be good too.

**Comment by Nilstrieb** [comment]
@onur-ozkan lol, thanks for mentioning the tests, I completely forgot about them and had to fix them

Another thing: I had to add this check:
https://github.com/rust-lang/rust/pull/121051/files#diff-b6fc213859d4da4176033b362381724914fea88a6c04122c0048ab5b76382343R1158

It was already kinda broken, lint-docs emitted warnings about things not quite matching up, so I don't think this should break anything and if it does it's good that it's broken.

**Comment by ehuss** [comment]
>  lint-docs emitted warnings about things not quite matching up, so I don't think this should break anything and if it does it's good that it's broken.

This behavior was intentional, and I'm a little concerned about changing it. We want to be able to build documentation with stage 0 (#73964 has some more context on that). The lint docs were designed to handle that case. Changing that causes `./x.py doc` to fail (with the default configuration). The doc command is designed to run in stage 0 by default.

Is it possible to have the target docs also just generate degraded output if using stage 0?

**Comment by ChrisDenton** [comment]
> Is it possible to have the target docs also just generate degraded output if using stage 0?

For now, we could just skip building target docs at `--stage 0`?

**Comment by oskgo** [comment]
@Noratrieb Any updates on this? Thanks

**Comment by JohnCSimon** [comment]
@Noratrieb 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this PR. 
Thanks 

@rustbot label: +S-inactive

**Pull request closed without merging**
